doc/man/polkit-check-caller.1.in
.\" 
.\" polkit-check-caller manual page.
.\" Copyright (C) 2007 David Zeuthen <david@fubar.dk>
.\"
.TH POLKIT-CHECK-CALLER 1
.SH NAME
polkit-check-caller \- check access
.SH SYNOPSIS
.PP
.B polkit-check-caller
[options]

.SH DESCRIPTION

\fIpolkit-check-caller\fP can be used to determine if a given caller
can a given action. For more information about the big picture refer
to the \fIPolicyKit spec\fP which can be found in
.I "@docdir@/spec/polkit-spec.html"
depending on the distribution.

.SH OPTIONS
The following options are supported:
.TP
.I "--action"
The action to check.
.TP
.I "--caller"
The caller to check for. Must be the callers unique name on the D-Bus
system message bus.
.TP
.I "--resource-type"
Type of resource.
.TP
.I "--resource"
Identifier of resource.
.TP
.I "--help"
Print out usage.
.TP
.I "--version"
Print the version.

.SH RETURN VALUE
.PP
If access is allowed, this
program exits with exit code 0. If no access is allowed or an error
occurs, the program exits with a non-zero exit code.

.SH BUGS
.PP
Please send bug reports to either the distribution or the HAL
mailing list, see 
.I "http://lists.freedesktop.org/mailman/listinfo/hal"
on how to subscribe.

.SH SEE ALSO
.PP
\&\fIPolicyKit\fR\|(8),
\&\fIdbus-daemon\fR\|(1),
\&\fIpolkit-check-session\fR\|(1)

.SH AUTHOR
Written by David Zeuthen <david@fubar.dk> with a lot of help from many
others.



NEWS
==========
PolicyKit 0.4 "Off the Grid"
==========

Released on July 25th 2007.

This is the second release of PolicyKit. 

WARNING WARNING WARNING: do not use this software in a stable
distribution; there are still security sensitive things that needs to
be audited.

Changes from PolicyKit 0.3 "No Man's Land"

David Zeuthen (18):
      post-release version bump
      remove the notion of modules
      replace configuration reload mechanism
      get proper pid and SELinux context
      add support for an /etc/PolicyKit/PolicyKit.conf config file
      document the /etc/PolicyKit/PolicyKit.conf file with a manual page
      update TODO
      switch from VIA_ROOT to VIA_ADMIN
      remove RESULT_NOT_AUTHORIZED_TO_KNOW and s/RESULT_UNKNOWN_ACTION/RESULT_UNKNOWN/
      fix up some of the docs
      change default username
      fix typo
      require that policy files also provide a <message> element
      fix docs
      fix up proper naming of some methods on the PolKitPolicyFileEntry class
      for PolKitContext, mention that a mechanism need to provide .policy files
      update NEWS
      actually reload config and policy files when they change

Requirements for PolicyKit 0.4 ""

 - dbus             >= 0.90
 - glib             >= 2.6.0
 - ConsoleKit       >= 0.2.1
 - expat            >= 1.95.8
 - libselinux       >= 1.30   (optional)

==========
PolicyKit 0.3 "No Man's Land"
==========

Released on June 20th 2007.

This is the first release of PolicyKit. 

WARNING WARNING WARNING: do not use this software in a stable
distribution; there are still security sensitive things that needs to
be audited.

Requirements for PolicyKit 0.3 "No Man's Land"

 - dbus             >= 0.90
 - glib             >= 2.6.0
 - ConsoleKit       >= 0.2.1
 - expat            >= 1.95.8



tools/polkit-check-caller.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*- */
/***************************************************************************
 *
 * polkit-check-caller.c : check if a caller is privileged
 *
 * Copyright (C) 2007 David Zeuthen, <david@fubar.dk>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307	 USA
 *
 **************************************************************************/

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <pwd.h>
#include <grp.h>
#include <unistd.h>
#include <errno.h>

#include <polkit-dbus/polkit-dbus.h>

#include <glib.h>

static void
usage (int argc, char *argv[])
{
	fprintf (stderr,
                 "\n"
                 "usage : polkit-check-caller\n"
                 "          --caller <dbus-name> --action <action>\n"
                 "          [--version] [--help]\n");
	fprintf (stderr,
                 "\n"
                 "        --caller         Unique name of caller on the system bus\n"
                 "        --action         Requested action\n"
                 "        --version        Show version and exit\n"
                 "        --help           Show this information and exit\n"
                 "\n"
                 "Determine if a given caller can do a given action. If access is \n"
                 "allowed, this program exits with exit code 0. If no access is allowed\n"
                 "or an error occurs, the program exits with a non-zero exit code.\n");
}

int
main (int argc, char *argv[])
{
        char *action_id = NULL;
        char *dbus_name = NULL;
        gboolean is_version = FALSE;
        DBusConnection *bus;
	DBusError error;
        PolKitContext *pol_ctx;
        PolKitCaller *caller;
        PolKitAction *action;
        gboolean allowed;
        PolKitError *p_error;
        
	if (argc <= 1) {
		usage (argc, argv);
		return 1;
	}

	while (1) {
		int c;
		int option_index = 0;
		const char *opt;
		static struct option long_options[] = {
			{"action", 1, NULL, 0},
			{"caller", 1, NULL, 0},
			{"version", 0, NULL, 0},
			{"help", 0, NULL, 0},
			{NULL, 0, NULL, 0}
		};

		c = getopt_long (argc, argv, "",
				 long_options, &option_index);
		if (c == -1)
			break;

		switch (c) {
		case 0:
			opt = long_options[option_index].name;

			if (strcmp (opt, "help") == 0) {
				usage (argc, argv);
				return 0;
			} else if (strcmp (opt, "version") == 0) {
				is_version = TRUE;
			} else if (strcmp (opt, "action") == 0) {
				action_id = strdup (optarg);
			} else if (strcmp (opt, "caller") == 0) {
				dbus_name = strdup (optarg);
			}
			break;

		default:
			usage (argc, argv);
			return 1;
			break;
		}
	}

	if (is_version) {
		printf ("polkit-check-caller " PACKAGE_VERSION "\n");
		return 0;
	}

	if (action_id == NULL || dbus_name == NULL) {
		usage (argc, argv);
		return 1;
	}

        dbus_error_init (&error);
        bus = dbus_bus_get (DBUS_BUS_SYSTEM, &error);
        if (bus == NULL) {
		fprintf (stderr, "error: dbus_bus_get(): %s: %s\n", error.name, error.message);
		return 1;
	}

        p_error = NULL;
        pol_ctx = polkit_context_new ();
        if (!polkit_context_init (pol_ctx, &p_error)) {
		fprintf (stderr, "error: polkit_context_init: %s\n", polkit_error_get_error_message (p_error));
                polkit_error_free (p_error);
                return 1;
        }

        action = polkit_action_new ();
        polkit_action_set_action_id (action, action_id);

        caller = polkit_caller_new_from_dbus_name (bus, dbus_name, &error);
        if (caller == NULL) {
                if (dbus_error_is_set (&error)) {
                        fprintf (stderr, "error: polkit_caller_new_from_dbus_name(): %s: %s\n", 
                                 error.name, error.message);
                        return 1;
                }
        }

        allowed = polkit_context_can_caller_do_action (pol_ctx, action, caller);

        if (allowed)
                return 0;
        else
                return 1;
}
